File: JDK11/opentraceability/src/main/java/opentraceability/utility/limitedpool`1.java
package opentraceability.utility;

import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.TimeUnit;

public class LimitedPool<T> {
    private ConcurrentLinkedDeque<LimitedPoolItem<T>> pool = new ConcurrentLinkedDeque<>();
    private boolean disposed = false;
    private long valueLifetime;
    private ValueFactory<T> valueFactory;
    private ValueDisposeAction<T> valueDisposeAction;

    public LimitedPool(ValueFactory<T> valueFactory, ValueDisposeAction<T> valueDisposeAction, long valueLifetime) {
        this.valueFactory = valueFactory;
        this.valueDisposeAction = valueDisposeAction;
        this.valueLifetime = valueLifetime;
    }

    public LimitedPool(ValueFactory<T> valueFactory, ValueDisposeAction<T> valueDisposeAction) {
        this(valueFactory, valueDisposeAction, TimeUnit.HOURS.toMillis(1));
    }

    public int getIdleCount() {
        return pool.size();
    }

    public LimitedPoolItem<T> get() {
        LimitedPoolItem<T> item = null;
        while (!disposed) {
            item = pool.pollLast();
            if (item == null || item.isExpired()) {
                if (item != null) {
                    item.dispose();
                }
            } else {
                return item;
            }
        }
        return new LimitedPoolItem<>(valueFactory.create(), (disposedItem) -> {
            if (disposedItem.isExpired()) {
                valueDisposeAction.dispose(disposedItem.getValue());
            } else {
                if (!disposed) {
                    pool.addLast(disposedItem);
                }
            }
        }, valueLifetime);
    }

    public void dispose() {
        disposed = true;
        LimitedPoolItem[] items = pool.toArray(new LimitedPoolItem[0]);
        for (LimitedPoolItem<T> item : items) {
            valueDisposeAction.dispose(item.getValue());
        }
    }

    public interface ValueFactory<T> {
        T create();
    }

    public interface ValueDisposeAction<T> {
        void dispose(T value);
    }
}

File: Java/OpenTraceability/src/main/opentraceability/utility/limitedpool`1.kt
package utility

import java.util.concurrent.ConcurrentStack
import kotlin.concurrent.getOrSet

class LimitedPool<T>(
    private val valueFactory: () -> T,
    private val valueDisposeAction: (T) -> Unit,
    valueLifetime: Long = 60 * 60 * 1000 // Default value lifetime of 1 hour
) : AutoCloseable {
    private val pool = ConcurrentStack<LimitedPoolItem<T>>()
    private var disposed = false

    val idleCount: Int
        get() = pool.size

    init {
        require(valueLifetime > 0) { "Value lifetime must be positive." }
    }

    fun get(): LimitedPoolItem<T> {
        var item: LimitedPoolItem<T>?
        while (!disposed && pool.tryPop().also { item = it } != null) {
            if (!item!!.expired) {
                return item!!
            }
            item!!.dispose()
        }
        return LimitedPoolItem(valueFactory(), { disposedItem ->
            if (disposedItem.expired) {
                valueDisposeAction(disposedItem.value)
            } else if (!disposed) {
                pool.push(disposedItem)
            }
        }, valueLifetime)
    }

    override fun close() {
        dispose()
    }

    fun dispose() {
        if (disposed) {
            return
        }
        disposed = true
        val items = pool.toTypedArray()
        items.forEach { valueDisposeAction(it.value) }
        pool.clear()
    }
}
